feat(docs): Add LLM-friendly markdown endpoints for AI agents

aryasaatvik opened the pull request:
## Summary

Adds AI-friendly markdown support to the SST documentation site, enabling AI agents and LLMs to access clean markdown versions of docs through multiple methods.

Inspired by [fumadocs PR #2408](https://github.com/fuma-nama/fumadocs/pull/2408) and [better-auth PR #5105](https://github.com/better-auth/better-auth/pull/5105), but implemented using Astro's native capabilities instead of HTML-to-markdown conversion.

## Features

### 1. Bulk Markdown Endpoints
- `/docs/llms.txt` - All documentation in a single file (~1.3MB)
- `/docs/llms-full.txt` - Alternative with table of contents

### 2. Direct `.md` URLs
Individual pages accessible via `.md` extension:
```bash
curl https://sst.dev/docs/components.md
curl https://sst.dev/blog/auth.md
```

### 3. HTTP Accept Header Negotiation
Middleware automatically serves markdown when AI agents prefer it:
```bash
curl -H "Accept: text/markdown" https://sst.dev/docs/components
```

### 4. Copy Markdown Button
User-facing feature: "Copy Markdown" button on all doc pages
- Right-aligned next to page title
- Icon state switching (clipboard → check)
- 2-second "Copied!" feedback
- Responsive design

## Implementation

- **Middleware** (`src/middleware.ts`) - Intercepts Accept headers and rewrites to `.md` endpoints
- **Dynamic routes** - Generate static `.md` files at build time for all docs/blog
- **Markdown cleaning** (`src/utils/markdown-clean.ts`) - Strips Astro components while preserving markdown structure
- **Bulk endpoints** - Serve all docs via Astro Content Collections API
- **UI component** - Copy button integrated into PageTitle component

## Architecture Benefits

**No HTML→Markdown conversion** - Direct access to source markdown  
**Static generation** - All routes pre-rendered at build time (no runtime overhead)  
**Astro-native** - Uses Content Collections API, no external dependencies  
**Backwards compatible** - Zero impact on existing HTML docs  
**AI-optimized** - Clean markdown without framework artifacts  

## Testing

```bash
# Bulk endpoints
curl https://sst.dev/docs/llms.txt | head -50

# Direct .md URLs
curl https://sst.dev/docs/components.md

# Accept header negotiation
curl -H "Accept: text/markdown" https://sst.dev/docs/components

# Copy button - visit any doc page in browser
```

## Related

- Fixes TypeDoc build error in `platform/src/components/aws/permission.ts` (missing JSDoc)
- Enables AI agents (Claude, ChatGPT, etc.) to consume SST docs natively
- Follows [llmstxt.org](https://llmstxt.org/) specification

aryasaatvik committed:
fix(platform): add missing JSDoc to permission.ts for TypeDoc

Adds module-level documentation with @packageDocumentation tag to fix
TypeDoc generation error: "Class comment not found".

This was blocking the build process during doc generation.

aryasaatvik committed:
feat(docs): add LLM-friendly markdown endpoints for AI agents

Implements multiple ways for AI agents and LLMs to access clean markdown documentation:

- Bulk endpoints at /docs/llms.txt and /docs/llms-full.txt serving all docs
- Direct .md URLs for individual pages (e.g., /docs/components.md)
- HTTP Accept header negotiation (text/markdown) with automatic rewrites
- Markdown cleaning utilities to strip Astro components and preserve structure

Architecture:
- Middleware intercepts Accept: text/markdown headers
- Dynamic routes generate static .md files at build time
- Content Collections API provides native access to raw markdown
- No HTML-to-markdown conversion needed (unlike fumadocs approach)

Enables AI-native documentation consumption without affecting human users.

aryasaatvik committed:
feat(docs): add copy markdown button to page titles

Adds a "Copy Markdown" button next to all documentation and blog page titles.

Features:
- Fetches markdown from .md endpoint on click
- Copies to clipboard using Navigator API
- Icon toggles between clipboard and check icons
- Shows "Copied!" feedback for 2 seconds
- Responsive layout (stacks vertically on mobile)
- Minimal, clean styling matching existing design

Button positioned right-aligned for easy access without disrupting title layout.

aryasaatvik committed:
refactor(docs): consolidate blog and docs .md routes

Consolidates separate blog/[...slug].md.ts and docs/[...slug].md.ts
into a single [section]/[...slug].md.ts route using dynamic section parameter.

Reduces code duplication and makes it easier to add new sections in the future.

Also includes linter formatting fixes.

aryasaatvik committed:
fix(docs): add static routes for index.md pages

Creates dedicated static routes for /docs/index.md and /blog/index.md
to work around Starlight's index route taking precedence over dynamic routes.

Also fixes slug lookup - Astro index files use parent directory as slug
(e.g., docs/index.mdx -> slug: "docs" not "docs/index").

aryasaatvik committed:
feat(docs): support index pages in markdown Accept header negotiation

Updates middleware to rewrite index pages (e.g., /docs/) to index.md
instead of skipping them. This enables full Accept header negotiation
for all docs and blog pages including index pages.

Rewrites:
- /docs/ -> /docs/index.md
- /docs/components -> /docs/components.md

aryasaatvik committed:
fix(docs): support /docs.md and /blog.md routes

Adds middleware rewrite to handle direct requests to /docs.md and /blog.md,
routing them to their respective index.md pages.

This ensures all these URLs work:
- /docs -> /docs/index.md (with Accept: text/markdown)
- /docs/ -> /docs/index.md (with Accept: text/markdown)
- /docs.md -> /docs/index.md (always)

jayair commented:
This is great, let me take a look.

jayair commented:
<img width="1346" height="174" alt="Screenshot 2025-10-13 at 7 35 40 PM" src="https://github.com/user-attachments/assets/563248cd-951a-4a40-a094-c4ea1b15aa17" />

I'm seeing this when I hit the copy markdown button in Safari.

jayair commented:
I also see the `/blog/*` in the `llms.txt`. Can we just restrict it to `/docs/*` maybe?

aryasaatvik committed:
fix(docs): filter out blog slugs from documentation listings

Updates the documentation retrieval logic to exclude any documents with slugs that start with "blog/", ensuring that only relevant documentation is sorted and returned. This improves the clarity and relevance of the documentation presented to users.

aryasaatvik committed:
fix(docs): execute navigator.clipboard.writeText in user context for safari

aryasaatvik commented:
> <img alt="Screenshot 2025-10-13 at 7 35 40 PM" width="1346" height="174" src="https://private-user-images.githubusercontent.com/53023/500697806-563248cd-951a-4a40-a094-c4ea1b15aa17.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA0MzQ1NDMsIm5iZiI6MTc2MDQzNDI0MywicGF0aCI6Ii81MzAyMy81MDA2OTc4MDYtNTYzMjQ4Y2QtOTUxYS00YTQwLWEwOTQtYzRlYTFiMTVhYTE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE0VDA5MzA0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyZTNkZWVkNzEzMzlmNzM2MGI5OGMwZDczYzJmZDNmMzUwNzhlNzQ4N2MyMTBhZDdlOWI2NzViMDVkNjc2ODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3tiuy_oOFUkl2mnKcv05W7abMJ4lBJV1bYvVh2qPrxY">
> I'm seeing this when I hit the copy markdown button in Safari. 

safari requires https context for navigator.clipboard.writeText https://webkit.org/blog/10247/new-webkit-features-in-safari-13-1/ . you can test this locally by adding the mkcert plugin `import mkcert from "vite-plugin-mkcert"` 

wrapped it in setTimeout because it navigator.clipboard needs to be called from the user context and the fetch before writeText break user context

vimtor reviewed (changes requested):
thanks for your contribution @aryasaatvik 

i'm trying the PR and found a couple of issues:

1. the "Copy Markdown" button doesn't look quite right (i'm on Chrome)

<img width="2738" height="1110" alt="Image" src="https://github.com/user-attachments/assets/96473752-4692-4583-9d14-8b1739fade7d" />

2. when fetching a page using the markdown accept header i still get the HTML version

<img width="3456" height="944" alt="Image" src="https://github.com/user-attachments/assets/b0fdced0-954c-4a1c-b93c-d032b6709d0d" />

vimtor commented:
closing in favour of https://github.com/anomalyco/sst/pull/6408
